plugins/modules/storage/zfs/zfs.py

```diff
@@ -186,10 +186,15 @@ def set_property(self, prop, value):
             self.module.fail_json(msg=err)
 
     def set_properties_if_changed(self):
+        diff = {'before': {}, 'after': {}}
         current_properties = self.get_current_properties()
         for prop, value in self.properties.items():
-            if current_properties.get(prop, None) != value:
+            current_value = current_properties.get(prop, None)
+            if current_value != value:
                 self.set_property(prop, value)
+                diff['before'][prop] = current_value
+                diff['after'][prop] = value
```

Comment:
This is a corner case, but one thing to consider is the behavior when the pool was imported/created with `-R` (or `-o altroot`). The result of `zfs-get` will not include the `altroot` value, so it will seem like the value has been changed when it really hasn't.

An example might make be a better way to explain:
``` shell
mkdir /tmp/fakepool
fallocate -l 256M /tmp/fakepool/vdev
zpool create -R /tmp/fakepool/mnt fakepool -O mountpoint=/ /tmp/fakepool/vdev
zfs get mountpoint fakepool # will return /tmp/fakepool/mnt
zfs set mountpoint=/ fakepool
zfs get mountpoint fakepool # will still return /tmp/fakepool/mnt
```
So my suggestion is that instead of setting `diff['after'][prop] = value`, the `diff['after]` values should be determined in the same way that `diff['before']` is (i.e. by calling `zfs-get`).

That will still lead to extra `zfs-set` calls being made (which isn't a big deal in my opinion), but it won't lead to the step being marked `changed` when it actually wasn't (which is more of an issue, especially if other steps are conditional on the `changed` result).

Reply:
Another situation where this might occur is with options that take sizes as parameters. Imagine that we have a pool `rpool` and run this command:
``` shell
zfs create -o quota=1G rpool/foobar
zfs get quota rpool/foobar # will return "1G"
zfs set quota=1024M rpool/foobar
zfs get quota rpool/foobar # will still return "1G"
```
So your current logic will consider that `zfs-set` call to have changed a value, even though nothing actually changed. I would say that calculating the `after` value with a second call to `zfs-get` is important for correctness.

Reply:
@sam-lunt 
as it's open on #975 we need to use '-p' on get values:
```
# zfs create -o quota=1G tank/q
# zfs get quota tank/q
NAME    PROPERTY  VALUE  SOURCE
tank/q  quota     1G     local
# zfs get -H -p quota tank/q
tank/q  quota   1073741824      local

# zfs set quota=1024M tank/q
# zfs get -H -p quota tank/q
tank/q  quota   1073741824      local
```

Reply:
Thanks, good point. Just to be clear though, I believe my point still stands. If the script compares the previous value (`1073741824`) with the user-provided value (`1024M`), then the task will be incorrectly marked as "changed". Using the `-p` certainly helps, but the real fix is comparing the previous output of `zfs-get` against the output of `zfs-get` after running `zfs-set` instead of against the values supplied in the playbook.

Reply:
In my opinion the diff mode should support the check mode and the apply mode (my word, hope you know, what I mean) and most importantly should generate the same diff output in both cases wherever possible. And as there are no actual changes in check mode, a comparison of the zfs get result before and after a "change" in such a situation wouldn't produce any diff output at all. Therefore I would rather vote against this approach and stick to the one proposed in this pull request. But please give me some time to dig into this topic again and complete my tests.

Reply:
I would suggest using a test similar to below:
``` yaml
- name: Set ZFS quota to 1G three times
  hosts: all
  tasks:
    - name: Set ZFS quota
      community.general.zfs:
        name: rpool/foobar
        state: present
        extra_zfs_properties:
          quota: "{{ item }}"
      loop:
        - "1G"
        - "1024M"
        - "1G"
```

With your current logic, either the 2nd or 3rd iteration of the loop will always be marked as "changed". I would say that neither should ever be considered "changed". I would say ensuring this is more important than supporting a "check" mode.

Reply:
The issues described above have already been addressed in PR #2454, which has been merged on 10.05.21. Therefore I consider this conversation as resolved.

Reply:
You are ignoring the fix that was introduced by PR #2454. When not running in check mode, you should use `update_properties` (set below) to calculate the `diff` instead of `self.properties`. As it is, you are going to have discrepancies where `self.changed` is false (correctly) but the `diff` shows properties as having been changed. (For example, something like `diff["before"]["quota"] = 1024` and `diff["after"]["quota"] = "1K"`)

Reply:
Good to know you on board, four eyes usually see more than two. I didn't ignore PR #2454. I actually thought quite a while about how to best integrate my changes into yours, but perhaps not thoroughly enough. My thought was, if the first comparison predicts changes but the second one doesn't detects actual ones, then the diff structure isn't displayed anyway during task execution, hence there is no real benefit of maintaining it after a change. But I've to admit, that this logic only applies well, if you only change one zfs property. If you change more than one, there might be partial changes. In these cases the diff structure might be displayed during task execution, even though not all properties were actually changed. Therefore I added some more code to address this. Please have a look at it and perhaps you could also put it to a test. I would really appreciate that. If you need test cases, I can describe some, but I've got the feeling, that you're quite familiar with this topic and have some interesting views on it, which I don't want to distract you from.